config.org
* Use-Package
#+BEGIN_SRC emacs-lisp
  ;; Make sure use package is available.
  (unless (package-installed-p 'use-package)
    (package-refresh-contents)
    (package-install 'use-package))

  ;; Configure use-package prior to loading it.
  (eval-and-compile
    (setq use-package-always-ensure nil)
    (setq use-package-always-defer nil)
    (setq use-package-always-demand nil)
    (setq use-package-enable-imenu-support t)
    ;; use real name of hook instead of shorter version.
    (setq use-package-hook-name-suffix nil))

  (eval-when-compile
    (require 'use-package))
#+END_SRC

* Set up variables
#+BEGIN_SRC emacs-lisp
  (use-package emacs
    :config
    (load-file (expand-file-name "~/.emacs.d/config-local-settings.el")))
#+END_SRC
* Appearance
** Disable GUI components
#+BEGIN_SRC emacs-lisp
  (use-package emacs
    :init
    (tool-bar-mode -1)
    (menu-bar-mode -1)
    (scroll-bar-mode -1)
    :config
    (setq use-file-dialog nil)
    (setq use-dialog-box t)
    (setq inhibit-startup-message t)
    :bind
    (("C-z" . nil)
     ("C-c C-z" . nil)
     ("C-h h" . nil)))
#+END_SRC
** Theme
*** manoj-dark
#+BEGIN_SRC emacs-lisp
  (use-package emacs
    :disabled
    :config
    (load-theme 'manoj-dark t))
#+END_SRC
*** sanityinc-tomorrow-blue
#+BEGIN_SRC emacs-lisp
  (use-package color-theme-sanityinc-tomorrow
    :ensure t
    :defer 0.1
    :config
    (load-theme 'sanityinc-tomorrow-blue t))
#+END_SRC
** Modeline
#+BEGIN_SRC emacs-lisp
  (use-package emacs
    :config
    (setq-default mode-line-format (list
                                    ;; saved, readonly
                                    "%*"
                                    ;; major mode
                                    "%m: "
                                    ;; buffer name
                                    "%b "
                                    ;; position of point
                                    "(%l,%c) ")))
#+END_SRC
** Fonts
#+BEGIN_SRC emacs-lisp
  (use-package frame
    :config
    (let* ((default-font-size (font-get (face-attribute 'default :font) :height))
           (default-font-family (font-get (face-attribute 'default :font) :family))
           (font-name (concat (if jacob-font-family
                                  (format "%s" jacob-font-family)
                                (format "%s" default-font-family))
                              "-"
                              (if jacob-font-size
                                  (format "%s" jacob-font-size)
                                (format "%s" default-font-size)))))
      (add-to-list 'default-frame-alist `(font . ,font-name))))
#+END_SRC
** beacon
 #+BEGIN_SRC emacs-lisp
   (use-package beacon
     :ensure t
     :demand
     :config
     (beacon-mode 1))
 #+END_SRC
** dimmer
#+BEGIN_SRC emacs-lisp
  (use-package dimmer
	:ensure t
    :defer 5
	:config
	(dimmer-mode))
#+END_SRC
** default frame size on startup
#+BEGIN_SRC emacs-lisp
  (use-package emacs
    :config
    (if (display-graphic-p)
        (setq initial-frame-alist
              '(
                (tool-bar-lines . 0)
                (width . 116)
                (fullscreen . fullheight)
                (left . 0)
                (top . 0)))))
#+END_SRC
* Built-in settings
** keyboard macros
#+BEGIN_SRC emacs-lisp
  (use-package kmacro
    :defer 0.1
    :after xah-fly-keys
    :bind
    (:map xah-fly-r-keymap
	  ("c" . kmacro-set-counter)))
#+END_SRC
** dabbrev
#+BEGIN_SRC emacs-lisp
  (use-package dabbrev
    :config
    (setq dabbrev-case-fold-search nil)
    (setq dabbrev-case-replace nil)
    :bind (("M-/" . dabbrev-expand)
           ("C-M-/" . dabbrev-completion)))
#+END_SRC
** put custom stuff in custom.el
#+BEGIN_SRC emacs-lisp
  (use-package cus-edit
    :config
    (defvar prot/custom-file "~/.emacs.d/custom.el")
    (setq custom-file prot/custom-file)
    (defun prot/cus-edit ()
      (let ((custom-file prot/custom-file))
        (unless (file-exists-p custom-file)
          (make-empty-file custom-file))
        (load-file custom-file))))
#+END_SRC
** prevent emacs from disabling commands.
#+BEGIN_SRC emacs-lisp
  (use-package novice
    :defer 0.1
    :config
    (setq disabled-command-function nil))
#+END_SRC
** move by logical lines
#+BEGIN_SRC emacs-lisp
  (use-package simple
    :defer 0.1
    :config
    (setq line-move-visual t))
#+END_SRC
** Use ls implemented in elisp to have consistent behaviour across platforms
#+BEGIN_SRC emacs-lisp
  (setq ls-lisp-use-insert-directory-program nil)
  (setq ls-lisp-dirs-first t)
#+END_SRC
** Read more data from process
#+BEGIN_SRC emacs-lisp
  (setq read-process-output-max (* 1024 1024))
#+END_SRC
** Reduce lag
#+BEGIN_SRC emacs-lisp
  (setq auto-window-vscroll nil)
  (setq redisplay-dont-pause t)
#+END_SRC
** Stop the bell ringing all the time
#+BEGIN_SRC emacs-lisp
  (setq ring-bell-function 'ignore)
#+END_SRC
** Prevent the view of the screen jumping to the middle when scrolling out of the view.
#+BEGIN_SRC emacs-lisp
  (setq scroll-conservatively 100)
#+END_SRC
** subword-mode
#+BEGIN_SRC emacs-lisp
  (use-package subword
    :defer 0.1
    :config (global-subword-mode))
#+END_SRC

** Make typing delete/overwrite region
#+BEGIN_SRC emacs-lisp
  (use-package delsel
    :defer 0.1
    :config
    (delete-selection-mode 1))
#+END_SRC

** Turn on bracket match highlight
#+BEGIN_SRC emacs-lisp
  (use-package paren
    :defer 0.1
    :config
    (show-paren-mode 1))
#+END_SRC
** Prevent emacs from creating debris when editing files
#+BEGIN_SRC emacs-lisp
  (use-package emacs
    :config
    (setq make-backup-files nil)
    (setq auto-save-default nil)
    (setq create-lockfiles nil))
#+END_SRC
** Backups don't destroy original file's creation date
#+BEGIN_SRC emacs-lisp
  (setq backup-by-copying t)
#+END_SRC

** Y/n instead of yes/no
#+BEGIN_SRC emacs-lisp
  (defalias 'yes-or-no-p 'y-or-n-p)
#+END_SRC

** Expert mode for ibuffer
#+BEGIN_SRC emacs-lisp
  (setq ibuffer-expert t)
#+END_SRC
** Add lisp folder to load path
#+BEGIN_SRC emacs-lisp
  (add-to-list 'load-path "~/.emacs.d/lisp/")
#+END_SRC
** Save minibuffer history (useful for compile)
#+BEGIN_SRC emacs-lisp
  (use-package savehist
    :defer 0.1
    :config
    (setq savehist-file "~/.emacs.d/savehist")
    (setq history-length 1000)
    (setq history-delete-duplicates t)
    (setq savehist-save-minibuffer-history t))
#+END_SRC
** Save cursor position in files
#+BEGIN_SRC emacs-lisp
  (use-package saveplace
    :config
    (setq save-place-file "~/.emacs.d/saveplace")
    (setq save-place-forget-unreadable-files t)
    (save-place-mode 1))
#+END_SRC
** Startup time
#+BEGIN_SRC emacs-lisp
(add-hook 'emacs-startup-hook
          (lambda ()
            (message "Emacs ready in %s with %d garbage collections."
                     (format "%.2f seconds"
                             (float-time
                              (time-subtract after-init-time before-init-time)))
                     gcs-done)))
#+END_SRC
** Microsoft Windows functions
#+BEGIN_SRC emacs-lisp
  (if (string-equal system-type "windows-nt")
      (use-package emacs
        :config
        (setq w32-pass-rwindow-to-system nil)
        (setq w32-pass-apps-to-system nil)
        (setq w32-rwindow-modifier 'super)
        (setq w32-apps-modifier 'hyper)))
#+END_SRC
** Tabs
#+BEGIN_SRC emacs-lisp
  (use-package emacs
    :config
    ;; use spaces to indent
    (setq-default indent-tabs-mode nil)
    ;; set default tab char's display width to 4 spaces
    (setq-default tab-width 4)
    ;; make tab key call indent command or insert tab character, depending on cursor position
    (setq-default tab-always-indent 'complete))
#+END_SRC
* Personal Functions
#+BEGIN_SRC emacs-lisp
  (defun jacob-original-find-file ()
    "Uses the original file-file mechanism. 
  Useful for dealing with files on other servers.
  (at least on Microsoft Windows)"
    (interactive)
    (let ((completing-read-function 'completing-read-default))
      (call-interactively 'find-file)))

  (defun eval-and-replace ()
    "Replace the preceding sexp with its value."
    (interactive)
    (backward-kill-sexp)
    (condition-case nil
        (prin1 (eval (read (current-kill 0)))
               (current-buffer))
      (error (message "Invalid expression")
             (insert (current-kill 0)))))

  (defun jacob-quit-popup-window ()
    (interactive)
    (let ((loop-list (window-list))
          (window-not-found t))
      (while (and loop-list window-not-found)
        (let* ((window (car loop-list))
               (mode (jacob-buffer-mode (window-buffer window))))
          (if (or (eq mode 'help-mode)
                  (eq mode 'compilation-mode)
                  (eq mode 'special-mode))
              (progn
                (quit-window :window window)
                (setq window-found nil))))
        (setq loop-list (cdr loop-list)))))

  (defun jacob-buffer-mode (buffer-or-string)
    "Returns the major mode associated with a buffer."
    (with-current-buffer buffer-or-string
       major-mode))

  (defun jacob-config-visit ()
    (interactive)
    (find-file "~/.emacs.d/config.org"))

  (defun jacob-config-reload ()
    (interactive)
    (org-babel-load-file (expand-file-name "~/.emacs.d/config.org")))

  (defun jacob-org-src-block ()
    "Replacement for C-c ' in both \"org-mode\" and when editing code blocks within \"org-mode\"."
    (interactive)
    (if (bound-and-true-p org-src-mode)
        (org-edit-src-exit)
      (if (equal major-mode 'org-mode)
          (org-edit-special))))

  (defun jacob-recompile-packages ()
    "Recompile all packages."
    (interactive)
    (byte-recompile-directory package-user-dir nil 'force))

  (defun jacob-split-window-below-select-new ()
    "Splits current window vertically, then switch to new window."
    (interactive)
    (split-window-below)
    (other-window 1))


  (defun jacob-split-window-right-select-new ()
    "Splits current window horizontally, then switch to new window."
    (interactive)
    (split-window-right)
    (other-window 1))

  (load-file (expand-file-name "~/.emacs.d/myLisp/jacob-long-time.el"))

  (defun jacob-display-time ()
    "Display the current date and time in the echo area."
    (interactive)
    (message (concat (format-time-string "%A the %e")
                     (jacob-day-suffix (string-to-number (format-time-string "%e")))
                     (format-time-string " of %B, the year of our Lord %Y, ")
                     "at "
                     (jacob-long-time (string-to-number (format-time-string "%H"))
                                      (string-to-number (format-time-string "%M")))
                     ".")))
#+END_SRC

* Xah Fly Keys
#+BEGIN_SRC emacs-lisp
  (use-package xah-fly-keys
    :ensure t

    :demand

    :init
    (setq xah-fly-use-control-key nil)

    ;; This is a keyboard macro that enters insert mode, presses a backspace, then returns to command mode.
    ;; It's purpose is so I can bind "D" in command mode to whatever backspace does in any given buffer.
    (fset 'backspace
          [?f backspace home])

    (fset 'enter
          [return])

    (defun jacob-xah-command-binds ()
      "Set custom keys for xah-fly-keys keybindings."
      (interactive)
      (define-key xah-fly-key-map (kbd "a") 'counsel-M-x)
      (define-key xah-fly-key-map (kbd "s") 'enter)
      (define-key xah-fly-key-map (kbd "8") 'er/expand-region)
      (define-key xah-fly-key-map (kbd "4") 'jacob-split-window-below-select-new)
      ;; 1 can be rebound, is bound to a inferior version of expand region
      (define-key xah-fly-key-map (kbd "2") 'jacob-quit-popup-window))

    :config
    (load-file (expand-file-name "~/.emacs.d/myLisp/jacob-xah-modified-commands.el"))

    (define-prefix-command 'jacob-config-keymap)
    (xah-fly-keys-set-layout "qwerty")
    (xah-fly-keys 1)

history
;; -*- mode: emacs-lisp; coding: utf-8-unix -*-
;; Minibuffer history file, automatically generated by ‘savehist’.

(setq savehist-minibuffer-history-variables '(input-method-history yes-or-no-p-history buffer-name-history which-key-keymap-history compile-history extended-command-history info-lookup-history regexp-history dired-shell-command-history ido-file-history counsel-describe-symbol-history shell-command-history query-replace-history minibuffer-history file-name-history ivy-history swiper-history counsel-M-x-history))
(setq yes-or-no-p-history '("yes"))
(setq which-key-keymap-history '(#("wdired" 0 6 (ivy-index 1))))
(setq extended-command-history '("xah-fly-keys" "which-key-show-keymap" "which-key-show-top-level"))
(setq regexp-history '("todo"))
(setq dired-shell-command-history '("xreader" "mpv -sub-file harry-potter-and-the-chamber-of-secrets-yify-english.srt" "mpv"))
(setq counsel-describe-symbol-history '(#("^rainbow-mode" 0 13 (ivy-index 0)) #("^sort-li" 0 8 (ivy-index 0)) #("^package-ac" 0 11 (ivy-index 0)) #("^" 0 1 (ivy-index 9221)) #("^elpy-enable" 0 12 (ivy-index 0)) #("^occur" 0 6 (ivy-index 0)) #("^ivy-occ" 0 8 (ivy-index 0)) #("^view-f" 0 7 (ivy-index 0)) #("^ivy-pus" 0 8 (ivy-index 0)) #("^push" 0 5 (ivy-index 0)) #("^yank" 0 5 (ivy-index 4)) #("^finger" 0 7 (ivy-index 0)) #("^load-theme" 0 11 (ivy-index 0))))
(setq shell-command-history '("dotnet run" "echo test" "pdflatex TaskAnalysis.tex" "git push" "git commit -m \"assignment ready for submission 24th of may 2020\"" "git add -A" "pdflatex *.tex" "xreader *.pdf" "pdflatex *.tex" "xreader *.pdf" "pdflatex *.tex" "shutdown now &" "xreader *.pdf" "pdflatex *.tex" "xreader *.pdf" "pdflatex *.tex" "xreader *.pdf" "pdflatex *.tex" "xreader *.pdf" "pdflatex *.tex" "xreader *.pdf" "pdflatex *.tex" "xreader *.pdf" "pdflatex *.tex" "xreader *.pdf" "xreader TaskAnalysis.pdf"))
(setq query-replace-history '("linearVelocity" "input" "rotationVelocity" "input" "_input" "Keys" "texture" "sprite" "Texture2D" "Texture" "jorgen" "misaki" ".X" "." ".Y" "." "jorgen" "misaki" "Console" "Debug" "Y" "X" "misaki" "misaski" ".0f" ".0" "deactive" "active" "Colour" "Color" "3.
" "3." "3.\\n" "3." "3. " "3." "deferred" "defered" "." "," "\\\\" "/" "\\\\" "\\" "\\\\" "\\" "/" "\\" "\\" "/" "hour" "true-hour" "(number-to-string minute)" "minute" "(number-to-string minute)" "minute" "minute" "hour" "(number-to-string hour)" "hour" "jacob-day-suffix" "day-suffix" "learned" "learnt" "access" "accessed" "C\\#" "csharp" "URL" "url" "Cov\\'ea" "Covea" "YouTube" "youtube" "URL" "url" ".NET" "dotnet" "\\#" "#"))
(setq minibuffer-history '("w" "e" "w" "e" "q" "e" "q" "w" "e" "q" "e" "q" "_misakiPosition" "_misakiTexture" "_misakiPosition" "keyboardState" "colourData" "xah" "q" "smex" "w" "e" "23"))

config.el
(setq disabled-command-function nil)

(if (display-graphic-p)
    (setq initial-frame-alist
          '(
            (tool-bar-lines . 0)
            (width . 116)
            (fullscreen . fullheight)
            (left . 0)
            (top . 0))))

(setq line-move-visual t)

(setq ls-lisp-use-insert-directory-program nil)
(setq ls-lisp-dirs-first t)

(setq read-process-output-max (* 1024 1024))

(setq inhibit-startup-message t)

(setq auto-window-vscroll nil)
(setq redisplay-dont-pause t)

(setq ring-bell-function 'ignore)

(setq scroll-conservatively 100)

(use-package subword
  :diminish
  :hook (prog-mode . subword-mode))

(delete-selection-mode 1)

(show-paren-mode 1)

(setq make-backup-files nil)

(setq auto-save-default nil)

(setq create-lockfiles nil)

(setq backup-by-copying t)

(defalias 'yes-or-no-p 'y-or-n-p)

(setq ibuffer-expert t)

(add-to-list 'load-path "~/.emacs.d/lisp/")

(savehist-mode 1)

(add-hook 'emacs-startup-hook
          (lambda ()
            (message "Emacs ready in %s with %d garbage collections."
                     (format "%.2f seconds"
                             (float-time
                              (time-subtract after-init-time before-init-time)))
                     gcs-done)))

(setq w32-pass-rwindow-to-system nil
	      w32-rwindow-modifier 'super)

(setq w32-pass-apps-to-system nil)
(setq w32-apps-modifier 'hyper)

;; use spaces to indent
(setq-default indent-tabs-mode nil)
;; set default tab char's display width to 4 spaces
(setq-default tab-width 4)
;; make tab key call indent command or insert tab character, depending on cursor position
(setq-default tab-always-indent t)

(defun eval-and-replace ()
  "Replace the preceding sexp with its value."
  (interactive)
  (backward-kill-sexp)
  (condition-case nil
      (prin1 (eval (read (current-kill 0)))
             (current-buffer))
    (error (message "Invalid expression")
           (insert (current-kill 0)))))

(defun jacob-quit-popup-window ()
  (interactive)
  (let ((loop-list (window-list))
        (window-not-found t))
    (while (and loop-list window-not-found)
      (let* ((window (car loop-list))
             (mode (jacob-buffer-mode (window-buffer window))))
        (if (or (eq mode 'help-mode)
                (eq mode 'compilation-mode)
                (eq mode 'special-mode))
            (progn
              (quit-window :window window)
              (setq window-found nil))))
      (setq loop-list (cdr loop-list)))))

(defun jacob-buffer-mode (buffer-or-string)
  "Returns the major mode associated with a buffer."
  (with-current-buffer buffer-or-string
     major-mode))

(defun config-visit ()
  (interactive)
  (find-file "~/.emacs.d/config.org"))

(defun config-reload ()
  (interactive)
  (org-babel-load-file (expand-file-name "~/.emacs.d/config.org")))

(defun jacob-org-src-block ()
  "Replacement for C-c ' in both \"org-mode\" and when editing code blocks within \"org-mode\"."
  (interactive)
  (if (bound-and-true-p org-src-mode)
      (org-edit-src-exit)
    (if (equal major-mode 'org-mode)
        (org-edit-special))))

(defun jacob-recompile-packages ()
  "Recompile all packages."
  (interactive)
  (byte-recompile-directory package-user-dir nil 'force))

(defun jacob-split-window-below-select-new ()
  "Splits current window vertically, then switch to new window."
  (interactive)
  (split-window-below)
  (other-window 1))


(defun jacob-split-window-right-select-new ()
  "Splits current window horizontally, then switch to new window."
  (interactive)
  (split-window-right)
  (other-window 1))

(use-package xah-fly-keys
  :ensure t

  :demand

  :init
  (setq xah-fly-use-control-key nil)

  ;; This is a keyboard macro that enters insert mode, presses a backspace, then returns to command mode.
  ;; It's purpose is so I can bind "D" in command mode to whatever backspace does in any given buffer.
  (fset 'backspace
        [?f backspace home])

  (fset 'enter
        [return])

  (defun jacob-xah-command-binds ()
    "Set custom keys for xah-fly-keys keybindings."
    (interactive)
    (define-key xah-fly-key-map (kbd "a") 'counsel-M-x)
    (define-key xah-fly-key-map (kbd "s") 'enter)
    (define-key xah-fly-key-map (kbd "8") 'er/expand-region)
    (define-key xah-fly-key-map (kbd "4") 'jacob-split-window-below-select-new)
    ;; 1 can be rebound, is bound to a inferior version of expand region
    (define-key xah-fly-key-map (kbd "2") 'jacob-quit-popup-window)) 

  :config
  (load-file (expand-file-name "~/.emacs.d/myLisp/jacob-xah-modified-commands.el"))

  (define-prefix-command 'jacob-config-keymap)
  (xah-fly-keys-set-layout "qwerty")
  (xah-fly-keys 1)

  (add-hook 'xah-fly-command-mode-activate-hook 'jacob-xah-command-binds)
  (jacob-xah-command-binds) ;; call it on startup so binds are set without calling xah-fly-command-mode-activate first.

  (add-hook 'dired-mode-hook 'xah-fly-keys-off)
  (add-hook 'eww-mode-hook 'xah-fly-keys-off)
  (add-hook 'ibuffer-mode-hook 'xah-fly-keys-off)
  (add-hook 'custom-mode-hook 'xah-fly-keys-off)

  (key-chord-define xah-fly-key-map "fd" 'xah-fly-command-mode-activate)

  :bind
  (:map jacob-config-keymap
        ("r" . config-reload)
        ("R" . restart-emacs)
        ("e" . config-visit)
        ("c" . jacob-org-src-block)
        ("p" . jacob-recompile-packages)
        ("t" . jacob-long-time-toggle))
  (:map xah-fly-e-keymap
        ("k". jacob-xah-insert-paren)
        ("l". jacob-xah-insert-square-bracket)
        ("j". jacob-xah-insert-brace)
        ("u". jacob-xah-insert-ascii-double-quote)
        ("i". jacob-xah-insert-ascii-single-quote)
        ("m" . xah-insert-hyphen)
        ("," . xah-insert-low-line)
        ("." . jacob-insert-equals)
        ("/" . jacob-insert-plus)
        ("z" . jacob-insert-apostrophe)
        ("x" . jacob-insert-at)
        ("c" . jacob-insert-hash)
        ("d" . backspace)
        ("v" . jacob-insert-tilde))
  (:map xah-fly-dot-keymap
        ("c" . jacob-config-keymap))
  (:map xah-fly-leader-key-map
        ("4" . jacob-split-window-right-select-new))
  (:map xah-fly-w-keymap
        ("n" . eval-and-replace)))

(use-package lsp-mode
  :ensure t
  :hook
  ((java-mode python-mode) . lsp)
  (lsp-mode . lsp-enable-which-key-integration)
  :commands lsp
  :init
  (setq lsp-completion-enable-additional-text-edit nil)
  (setq lsp-prefer-capf nil)
  (setq lsp-prefer-flymake nil)
  :config
  (define-key xah-fly-dot-keymap (kbd "l") lsp-command-map))

(use-package lsp-ui
  :ensure t
  :commands lsp-ui-mode)

(use-package lsp-ivy
  :ensure t
  :commands lsp-ivy-workspace-symbol)

(use-package dap-mode
  :ensure t
  :hook java-mode
  :config
  (use-package dap-java)
  (dap-mode 1)
  (dap-ui-mode 1)
  (dap-tooltip-mode 1)
  (tooltip-mode 1)
  (dap-ui-controls-mode 1))

(use-package lsp-java
  :ensure t)

(use-package lsp-python-ms
  :ensure t
  :init (setq lsp-python-ms-auto-install-server t))

(use-package dired
  :config
  (defun jacob-teardown-xah-for-wdired ()
    (interactive)
    (wdired-finish-edit)
    (define-key xah-fly-leader-key-map (kbd ";") 'save-buffer)
    (xah-fly-keys-off))

  (defun jacob-setup-xah-for-wdired ()
    (interactive)
    (xah-fly-keys)
    (define-key xah-fly-leader-key-map (kbd ";") 'jacob-teardown-xah-for-wdired))

  (add-hook 'wdired-mode-hook 'jacob-setup-xah-for-wdired)

  (define-key dired-mode-map (kbd "RET") 'dired-find-alternate-file)
  (define-key dired-mode-map (kbd "^")(lambda () (interactive)(find-alternate-file "..")))
  (setq dired-dwim-target t)

  :bind
  (:map dired-mode-map
        ("," . switch-window)
        ("SPC" . xah-fly-leader-key-map)
        ("p" . dired-maybe-insert-subdir)
        ("i" . dired-previous-line)
        ("k" . dired-next-line)
        ("n" . isearch-forward)
        ("f" . dired-toggle-read-only)
        ("q" . xah-close-current-buffer)))

(use-package bnf-mode
  :ensure t)

(use-package org
  :mode ("\\.org\\'" . org-mode)
  :config
  (add-to-list 'org-structure-template-alist
             '("el" "#+BEGIN_SRC emacs-lisp\n?\n#+END_SRC")))

(use-package yaml-mode
  :ensure t
  :defer t
  :mode ("\\.yml\\'" . yaml-mode))

(setq-default c-basic-offset 4)

(use-package csharp-mode
  :ensure t
  :defer t
  :config
  (defun my-csharp-mode-setup ()
    (setq c-syntactic-indentation t)
    (c-set-style "ellemtel")
    (setq c-basic-offset 4)
    (load-file "~/.emacs.d/myLisp/namespace.el"))
  :hook
  (csharp-mode . my-csharp-mode-setup)
  :mode
  ("\\.cs\\$" . csharp-mode))

(use-package web-mode
  :ensure t

  :preface
  (defun jacob-web-mode-config ()
    (interactive)
    (setq-local electric-pair-pairs '((?\" . ?\") (?\< . ?\>)))
    (yas-activate-extra-mode 'html-mode))

  :config
  (setq web-mode-engines-alist
                '(("razor"	. "\\.cshtml\\'")))
  (setq web-mode-markup-indent-offset 2)
  (setq web-mode-css-indent-offset 2)
  (setq web-mode-code-indent-offset 2)

  :hook (web-mode . jacob-web-mode-config)

  :mode (("\\.html?\\'" . web-mode)
         ("\\.cshtml\\'" . web-mode)
         ("\\.css\\'" . web-mode)))

(use-package json-mode
  :ensure t
  :mode ("\\.json\\$" . json-mode))

(use-package clojure-mode
  :ensure t
  :mode ("\\.clj\\$" . clojure-mode))

(use-package flycheck
  :ensure t
  :defer 2
  :config
  (global-flycheck-mode))

(use-package beacon
  :ensure t
  :demand
  :diminish
  :config
  (setq beacon-color "#f2777a")
  (beacon-mode 1))

(use-package which-key
      :ensure t
  :defer 2
      :diminish
      :config
      (which-key-mode))

(use-package company
  :ensure t
  :defer t
  :diminish
  :hook ((emacs-lisp-mode csharp-mode java-mode) . company-mode)
  :config
  (setq company-idle-delay 0.5)
  (setq company-minimum-prefix-length 3))

(use-package projectile
  :ensure t
  :defer 2
  :diminish
  :config
  (projectile-mode t)
  (define-key xah-fly-dot-keymap (kbd "p") projectile-command-map)
  (setq projectile-completion-system 'ivy))

(use-package dimmer
      :ensure t
  :defer 5
      :config
      (dimmer-mode))

(use-package omnisharp
   :ensure t
   :defer t
   :after company
   :hook (csharp-mode . omnisharp-mode)
   :bind
   (:map jacob-omnisharp-keymap
         ("u" . omnisharp-fix-usings)
         ("U" . omnisharp-find-usages)
         ("i" . omnisharp-find-implementations)
         ("d" . omnisharp-go-to-definition)
         ("r" . omnisharp-rename)
         ("a" . omnisharp-run-code-action-refactoring)
         ("o" . omnisharp-start-omnisharp-server)
         ("O" . omnisharp-stop-server))
   :config
   (define-prefix-command 'jacob-omnisharp-keymap)
   (define-key xah-fly-dot-keymap (kbd "o") jacob-omnisharp-keymap)
   (add-hook 'omnisharp-mode-hook (lambda ()
                                    (add-to-list (make-local-variable 'company-backends)
                                                 '(company-omnisharp))))
   (setq omnisharp-company-ignore-case nil)
   (setq omnisharp-server-executable-path "D:\\Programming\\OmniSharp\\omnisharp-roslyn\\bin\\Debug\\OmniSharp.Stdio.Driver\\net472\\OmniSharp.exe"))

(use-package yasnippet
  :ensure t

  :hook
  (((csharp-mode web-mode python-mode java-mode) . yas-minor-mode))

  :config
  (yas-reload-all))

(use-package key-chord
  :defer 1

  :config
  (key-chord-mode 1))

(use-package cider
  :diminish
  :ensure t
  :mode ("\\.clj\\$" . clojure-mode))

(use-package avy
  :ensure t
  :defer 0.1
  :config
  (setq avy-keys (number-sequence ?a ?z))
  (setq avy-all-windows t)
  (setq avy-orders-alist
        '((avy-goto-char-timer . avy-order-closest)
          (avy-goto-end-of-line . avy-order-closest)))
  (key-chord-define xah-fly-key-map "fj" 'avy-goto-word-or-subword-1)
  (key-chord-define xah-fly-key-map "f;" 'avy-goto-end-of-line))

(use-package minibuffer-line
  :ensure t
  :init
  (minibuffer-line-mode)
  :config
  (load "~/.emacs.d/myLisp/jacob-long-time")

  (set-face-attribute 'minibuffer-line nil :foreground "#ffffff" :background "#002451")

  (setq-default minibuffer-line-format (list
                                ;; date
                                '(:eval (concat (format-time-string "%A the %e")
                                                (jacob-day-suffix (string-to-number (format-time-string "%e")))
                                                (format-time-string " of %B %Y, ")))
                                ;; time
                                '(:eval (concat "at "
                                                (jacob-long-time (string-to-number (format-time-string "%H")) (string-to-number (format-time-string "%M")))))))
  (face-remap-reset-base 'minibuffer-line)
  (minibuffer-line--update))

(use-package restart-emacs
      :ensure t
      :defer t)

(use-package smex
  :ensure t
  :config (smex-initialize)
  :bind
  ("M-x" . smex))

(use-package diminish
      :ensure t
      :defer t
      :config
      (diminish 'subword-mode)
      (diminish 'org-src-mode)
      (diminish 'eldoc-mode))

(use-package switch-window
      :ensure t
      :defer t
      :config
      (setq switch-window-input-style 'minibuffer)
      (setq switch-window-threshold 2)
      (setq switch-window-multiple-frames t)
      (setq switch-window-shortcut-style 'qwerty)
      (setq switch-window-qwerty-shortcuts
		'("q" "w" "e" "r" "a" "s" "d" "f" "z" "x" "c" "v"))
      :bind
      ([remap xah-next-window-or-frame] . switch-window))

(use-package ivy
  :ensure t
  :diminish
  :defer 0.1

  :bind
  (:map xah-fly-leader-key-map
        ("v" . counsel-yank-pop))

  :config
  (setq ivy-initial-inputs-alist nil)
  (setq enable-recursive-minibuffers t)
  (ivy-mode 1))

(use-package swiper
  :ensure t
  :after ivy
  :bind
  (:map xah-fly-dot-keymap
        ("s" . swiper)))

(use-package counsel
  :ensure t
  :diminish
  :after ivy

  :config (counsel-mode)
  :bind
  (:map xah-fly-c-keymap
        ("e" . counsel-find-file)))

(use-package multiple-cursors
      :ensure t
      :bind
      (:map xah-fly-dot-keymap
		("m" . jacob-multiple-cursors-keymap)
      :map jacob-multiple-cursors-keymap
		("l" . mc/edit-lines)
		(">" . mc/mark-next-like-this)
		("<" . mc/mark-previous-like-this)
		("a" . mc/mark-all-like-this))
      :init
      (define-prefix-command 'jacob-multiple-cursors-keymap))

(use-package expand-region
  :ensure t

  :config
  (setq expand-region-contract-fast-key "9"))

(use-package shell-pop
  :ensure t

  :config
  (setq shell-pop-autocd-to-working-dir nil)
  (setq shell-pop-shell-type (quote ("eshell" "*eshell*" (lambda nil (eshell)))))
  (setq shell-pop-universal-key "<H-return>")
  (setq shell-pop-window-position "bottom")
  (setq shell-pop-window-size 50)

  (defun jacob-shell-pop-eshell ()
    (interactive)
    (let ((shell-pop-shell-type '("eshell" "*eshell*" (lambda () (eshell))))
          (shell-pop-term-shell "eshell"))
      (shell-pop--set-shell-type 'shell-pop-shell-type shell-pop-shell-type)
      (call-interactively 'shell-pop)))

  (defun jacob-shell-pop-shell ()
    (interactive)
    (let ((shell-file-name "C:/Windows/System32/Cmd.exe")
          (shell-pop-shell-type '("shell" "*shell*" (lambda () (shell))))
          (shell-pop-term-shell "shell"))
      (shell-pop--set-shell-type 'shell-pop-shell-type shell-pop-shell-type)
      (call-interactively 'shell-pop)))

  :bind
  (:map xah-fly-n-keymap
        ("d" . jacob-shell-pop-eshell)
        ("f" . jacob-shell-pop-shell)))

(use-package eshell-up
      :ensure t)

(use-package langtool
      ;; :ensure t
      :defer t
      :config
      (setq langtool-language-tool-jar
		"/home/lem/Documents/LanguageTool-4.8/languagetool-commandline.jar"))

(use-package color-theme-sanityinc-tomorrow
  :ensure t
  :defer 0.1
  :config
  (load-theme 'sanityinc-tomorrow-blue t))

(setq-default mode-line-format (list
                                ;; saved, readonly
                                "%*"
                                ;; major mode
                                "%m: "
                                ;; buffer name
                                "%b "
                                ;; position of point
                                "(%l,%c) "))

(tool-bar-mode -1)

(menu-bar-mode -1)

(scroll-bar-mode -1)

(when (member "DejaVu Sans Mono" (font-family-list))
	(add-to-list 'initial-frame-alist '(font . "DejaVu Sans Mono-10"))
	(add-to-list 'default-frame-alist '(font . "DejaVu Sans Mono-10")))
